internal/logger/splunk_logger.go

```diff
@@ -0,0 +1,135 @@
+package logger
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+)
+
+const (
+	PayloadsChannelSize = 1000
+	// in seconds, how often batched events should be sent
+	SendFrequency = 5
+)
+
+type SplunkLogger struct {
+	client   *http.Client
+	url      string
+	token    string
+	source   string
+	hostname string
+
+	payloads chan *SplunkPayload
+}
+
+type SplunkPayload struct {
+	// splunk expects unix time in seconds
+	Time  int64       `json:"time"`
+	Host  string      `json:"host"`
+	Event SplunkEvent `json:"event"`
+}
+
+type SplunkEvent struct {
+	Message string `json:"message"`
+	Ident   string `json:"ident"`
+	Host    string `json:"host"`
+}
+
+func NewSplunkLogger(url, token, source, hostname string) *SplunkLogger {
+	sl := &SplunkLogger{
+		client:   &http.Client{},
+		url:      url,
+		token:    token,
+		source:   source,
+		hostname: hostname,
+	}
+
+	ticker := time.NewTicker(time.Second * SendFrequency)
+	sl.payloads = make(chan *SplunkPayload, PayloadsChannelSize)
+
+	go sl.flushPayloads(ticker.C)
+
+	return sl
+}
+
+func (sl *SplunkLogger) flushPayloads(ticker <-chan time.Time) {
+	var payloads []*SplunkPayload
+	for {
+		select {
+		case p := <-sl.payloads:
+			if p != nil {
+				payloads = append(payloads, p)
+			}
+			if len(payloads) == PayloadsChannelSize {
+				err := sl.SendPayloads(payloads)
+				if err != nil {
+					fmt.Fprintf(os.Stderr, "Splunk logger unable to send payloads: %v", err)
+				}
+				payloads = nil
+			}
+		case <-ticker:
+			err := sl.SendPayloads(payloads)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Splunk logger unable to send payloads: %v", err)
+			}
+			payloads = nil
+		}
+	}
+}
+
+func (sl *SplunkLogger) SendPayloads(payloads []*SplunkPayload) error {
+	if len(payloads) == 0 {
+		return nil
+	}
+
+	buf := bytes.NewBuffer(nil)
+	for _, pl := range payloads {
+		b, err := json.Marshal(pl)
+		if err != nil {
+			return err
+		}
+
+		_, err = buf.Write(b)
+		if err != nil {
+			return err
+		}
+	}
+
+	req, err := http.NewRequest("POST", sl.url, bytes.NewReader(buf.Bytes()))
+	if err != nil {
+		return err
+	}
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Authorization", fmt.Sprintf("Splunk %s", sl.token))
+
+	res, err := sl.client.Do(req)
+	if err != nil {
```

Comment:
I think it makes sense to retry once or twice for failed requests, may be this can be added later using retryablehttp.
It is not common but I believe it used  to happen occasionally, probably once every week or two.

Reply:
makes sense, opened https://github.com/osbuild/image-builder/issues/525 as a followup.